src/common/proc_macro/src/session_config.rs

```diff
@@ -0,0 +1,262 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use bae::FromAttributes;
+use proc_macro2::TokenStream;
+use proc_macro_error::{abort, OptionExt, ResultExt};
+use quote::{format_ident, quote, quote_spanned};
+use syn::DeriveInput;
+
+#[derive(FromAttributes)]
+pub struct Parameter {
+    pub rename: Option<syn::LitStr>,
+    pub default: syn::Expr,
+    pub flags: Option<syn::LitStr>,
+    pub check_hook: Option<syn::Expr>,
+}
+
+pub(crate) fn derive_config(input: DeriveInput) -> TokenStream {
+    let syn::Data::Struct(syn::DataStruct { fields, .. }) = input.data else {
+        abort!(input, "Only struct is supported");
+    };
+
+    let mut default_fields = vec![];
+    let mut struct_impl_set = vec![];
+    let mut struct_impl_get = vec![];
+    let mut struct_impl_reset = vec![];
+    let mut set_match_branches = vec![];
+    let mut get_match_branches = vec![];
+    let mut show_all_list = vec![];
+
+    for field in fields {
+        let field_ident = field.ident.expect_or_abort("Field need to be named");
+        let ty = field.ty;
+
+        let mut doc_list = vec![];
+        for attr in &field.attrs {
+            if attr.path.is_ident("doc") {
+                let meta = attr.parse_meta().expect_or_abort("Failed to parse meta");
+                if let syn::Meta::NameValue(val) = meta {
+                    if let syn::Lit::Str(desc) = val.lit {
+                        doc_list.push(desc.value().trim().to_string());
+                    }
+                }
+            }
+        }
+
+        let description: TokenStream = format!("r#\"{}\"#", doc_list.join(" ")).parse().unwrap();
+
+        let attr =
+            Parameter::from_attributes(&field.attrs).expect_or_abort("Failed to parse attribute");
+        let Parameter {
+            rename,
+            default,
+            flags,
+            check_hook: check_hook_name,
+        } = attr;
+
+        let entry_name = if let Some(rename) = rename {
+            if !(rename.value().is_ascii() && rename.value().to_ascii_uppercase() == rename.value())
+            {
+                abort!(rename, "Expect `rename` to be an ascii upper case string");
```

Comment:
Just a FYI: Actually PG seems to prefer lower-case style for session variables.

For example: https://www.postgresql.org/docs/current/config-setting.html#CONFIG-SETTING-NAMES-VALUES

Reply:
Maybe we can add an `allowed_value`?

Reply:
> PG seems to prefer lower-case style for session variables

It is actually case preserving yet insensitive (rust crate [uncased](https://docs.rs/uncased/latest/uncased/), postgresql src [guc_tables.c](https://github.com/postgres/postgres/blob/3edc6580c0e27fb8f13322efd255a88d20dda6c2/src/backend/utils/misc/guc_tables.c#L3957)):
```
test=# show datestyle;
 DateStyle 
-----------
 ISO, MDY
(1 row)
```

Reply:
Replied at https://github.com/risingwavelabs/risingwave/pull/13554#discussion_r1402957583